analyzers/src/SonarAnalyzer.CSharp/Rules/AspNet/UseAspNetModelBinding.cs

```diff
@@ -0,0 +1,284 @@
+                refKind: RefKind.None),
+            ArgumentDescriptor.MethodInvocation(// Request.Headers.ContainsKey("id")
+                invokedMemberConstraint: x => IsIDictionaryStringStringValuesInvocation(x, ContainsKey),
+                invokedMemberNameConstraint: (name, comparison) => string.Equals(name, ContainsKey, comparison),
+                invokedMemberNodeConstraint: IsAccessedViaHeaderDictionary,
+                parameterConstraint: x => string.Equals(x.Name, "key", StringComparison.Ordinal),
+                argumentListConstraint: (list, _) => list.Count == 1,
+                refKind: RefKind.None),
+            ArgumentDescriptor.ElementAccess(// Request.Query["id"]
+                invokedIndexerContainer: KnownType.Microsoft_AspNetCore_Http_IQueryCollection,
+                invokedIndexerExpression: "Query",
+                parameterConstraint: _ => true, // There is only a single overload and it is getter only
+                argumentPosition: 0),
+            ArgumentDescriptor.MethodInvocation(// Request.Query.TryGetValue("id", out _)
+                invokedType: KnownType.Microsoft_AspNetCore_Http_IQueryCollection,
+                methodName: TryGetValue,
+                parameterName: "key",
+                argumentPosition: 0),
+            ArgumentDescriptor.ElementAccess(// Request.RouteValues["id"]
+                invokedIndexerContainer: KnownType.Microsoft_AspNetCore_Routing_RouteValueDictionary,
+                invokedIndexerExpression: "RouteValues",
+                parameterConstraint: IsGetterParameter, // RouteValues are read/write
+                argumentPosition: 0),
+            ArgumentDescriptor.MethodInvocation(// Request.RouteValues.TryGetValue("id", out _)
+                invokedType: KnownType.Microsoft_AspNetCore_Routing_RouteValueDictionary,
+                methodName: TryGetValue,
+                parameterName: "key",
+                argumentPosition: 0)]);
+
+        propertyAccessDescriptors.Add(new(KnownType.Microsoft_AspNetCore_Http_IFormCollection, "Files")); // Request.Form.Files...
+    }
+
+    // Check that the "Headers" expression in the Headers.TryGetValue("id", out _) invocation is of type IHeaderDictionary
+    private static bool IsAccessedViaHeaderDictionary(SemanticModel model, ILanguageFacade language, SyntaxNode invocation) =>
+        invocation is InvocationExpressionSyntax { Expression: { } expression }
+            && GetLeftOfDot(expression) is { } left
+            && model.GetTypeInfo(left) is { Type: { } typeSymbol } && typeSymbol.Is(KnownType.Microsoft_AspNetCore_Http_IHeaderDictionary);
+
+    private static bool IsOverridingFilterMethods(ISymbol owningSymbol) =>
+        (owningSymbol.GetOverriddenMember() ?? owningSymbol).ExplicitOrImplicitInterfaceImplementations().Any(x => x is IMethodSymbol { ContainingType: { } container }
+            && container.IsAny(
+                KnownType.Microsoft_AspNetCore_Mvc_Filters_IActionFilter,
```

Comment:
Move these types to a `static readonly` field to avoid array allocation during the method call.